lsp-mode.el

```diff
@@ -6005,9 +5979,28 @@ returns the command to execute."
                        (cons proc proc))))
         :test? (lambda () (-> local-command lsp-resolve-final-function lsp-server-present?))))
 
+(defun lsp--setup-company ()
+  (company-mode 1)
+  (add-hook 'company-completion-started-hook
```

Comment:
Can all these hooks be performed with `(with-eval-afterload)`?
Reason: I'm not using `lsp--auto-configure` and have more sophisticated company backends configuration to handle capf.
Moreover, the hook should be added whether auto configured is set or not.

Reply:
Isn't `(add-hook 'lsp-managed-mode-hook 'lsp--setup-company)` sufficient to cover this case?

In general, executing stuff on the top level is considered a flaw so I want to not have it where possible.

Reply:
It's not a good idea to use `eval-after-load` in the package.

Reply:
Yes, the `add-hook` is enough to cover that case. However, are we asking the user to additionally add hook to a private `lsp--setup-company` function? That's kind of breaking change for user who don't use auto configure.
Instead of executing stuff on the top level, perhaps we can add the `with-eval-afterload` to minor mode initialization function?
The part of enable `(company-mode 1)` can stay in its places before this change though.

Reply:
Instead of usinig `eval-after-load` forward declaring hook and add hook immediately works for me too.

Reply:
> However, are we asking the user to additionally add hook to a private lsp--setup-company function?

 We can reorganize and make each of the functions from the auto configure public and reusable. E. g. lsp-setup-flycheck, lsp-setup-company so the user could pick only the functions that he/she wants. 

> Instead of usinig eval-after-load forward declaring hook and add hook immediately works for me too.

This means that we will set these hooks for all modes. Although our handlers are very lightweight it doesn't feel right to have them unconditionally added.

Reply:
> We can reorganize and make each of the functions from the auto configure public and reusable. E. g. lsp-setup-flycheck, lsp-setup-company so the user could pick only the functions that he/she wants.

That would be great. Albeit I still need to modify my config file for `lsp-mode` to work again. In case if user is not aware of those setup function, does that means `lsp-mode` is working incorrectly? Do we want that?

> > Instead of usinig eval-after-load forward declaring hook and add hook immediately works for me too.
> 
> This means that we will set these hooks for all modes. Although our handlers are very lightweight it doesn't feel right to have them unconditionally added.

I mean we can add hook immediately without waiting for the feature to be loaded. We can add hook in mode enable function and clean it up in mode disable function. That will ensure that the hook will only affect when `lsp-mode` is running.

Reply:
> I mean we can add hook immediately without waiting for the feature to be loaded. We can add hook in mode enable function and clean it up in mode disable function. That will ensure that the hook will only affect when `lsp-mode` is running.

This sounds ok - I will add them directly to the managed mode hook to make sure they are executed even with autoconfigure disabled. But generally, I consider disabling autoconfigure options to be only for the most advanced users/contributors which follow lsp-mode development.

Reply:
> But generally, I consider disabling autoconfigure options to be only for the most advanced users/contributors which follow lsp-mode development.

I see. I've tried `lsp-auto-configure` before in the hope of not getting burn if there're breaking change in `lsp-mode` that require my config change to get it works.
However I found out my `companymode` setup is become broken instead.

Should we add another custom `lsp-enable-company-integration` to help auto configure to automatically changing company mode configuration or not?

Reply:
> Should we add another custom `lsp-enable-company-integration` to help auto configure to automatically changing company mode configuration or not?

I don't want to go in that direction - I will follow your advice and use the managed mode hook unconditionally since we need this to make company work.

Reply:
I pushed the new version - moved the setup company in managed mode.